src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/Cas1ApplicationStatusService.kt

```diff
@@ -23,6 +27,38 @@ class Cas1ApplicationStatusService(
     bookingMade(spaceBooking.application!!)
   }
 
+  fun bookingCancelled(
+    booking: BookingEntity,
+    isUserRequestedWithdrawal: Boolean,
+  ) {
+    if (!isUserRequestedWithdrawal || booking.application == null) {
+      return
+    }
+    val application = booking.application!!
+    val bookings = bookingRepository.findAllByApplication(application)
+    val anyActiveBookings = bookings.any { it.isActive() }
+    if (!anyActiveBookings) {
+      bookingCancelled(booking.application!! as ApprovedPremisesApplicationEntity)
+    }
+  }
+
+  fun spaceBookingCancelled(spaceBooking: Cas1SpaceBookingEntity, isWithdrawableTreeOperation: Boolean = false) {
+    if (isWithdrawableTreeOperation || spaceBooking.application == null) {
+      return
+    }
+    val application = spaceBooking.application!!
+    val spaceBookingsForApplication = cas1SpaceBookingRepository.findAllByApplication(application)
+    val anyActiveBookings = spaceBookingsForApplication.any { it.isActive() }
+    if (!anyActiveBookings) {
+      bookingCancelled(spaceBooking.application!!)
+    }
+  }
```

Comment:
i _think_ we may have to return to this once we have a way to determine which premise works with space bookings and which doesn't because it's possible for an application to have a mix of bookings across regions (some supporting space bookings, some not). I'll make a note on that ticket